### ISSUE TITLE
SSN was saved as the file number on a veteran record

### ISSUE THREAD

**Issue by shanear** [department-of-veterans-affairs/caseflow#9298]
We saw an instance in prod where the SSN was saved as the file number in prod. We need to investigate why that happened, and prevent it from happening in the future.

It prevented the upload of a decision document. I fixed the data and resolved the decision document upload.

The veteran affected was:
```
Veteran id: 58874
```

## AC
We need to do 2 things: 
1. To prevent it from happening again:
   - we should try to fetch by file number first instead of assuming SSN is the file number.

## Out of scope - should do in another ticket
2. To clean-up existing Veteran records where this has already happened
   - investigate potential solutions for cleaning up every place where we have the Veteran File Number column?

**Comment by shanear** [comment]
Okay in order to fix this I had to delete the veteran. I also think this could have been due to a de-dup in corp DB. Worth a little bit more investigating

**Comment by shanear** [comment]
Update script:
```
old = "OLDFILENUMBER"
new = "NEWFILENUMBER"
dry_run = true

a = Appeal.where(veteran_file_number: old)
l = AvailableHearingLocations.where(veteran_file_number: old)
d = Document.where(file_number: old)
e = EndProductEstablishment.where(veteran_file_number: old)
i = Intake.where(veteran_file_number: old)
h = HigherLevelReview.where(veteran_file_number: old)
s = SupplementalClaim.where(veteran_file_number: old)
f = Form8.where(file_number: old)
r = RampElection.where(veteran_file_number: old)
rr = RampRefiling.where(veteran_file_number: old)
la = LegacyAppeal.where(vbms_id: old)
av = ApiView.where(vbms_id: old)

v = Veteran.where(file_number: old)
newv = Veteran.where(file_number: new).last
vthing = newv ? :delete : :update

if !dry_run
  [a, l, e, i, h, s, r, rr].each do |records|
    fail "Something looks wrong" if records.count > 1000

    records.update_all(veteran_file_number: new)
  end

  [d, f].each do |records|
    fail "Something looks wrong" if records.count > 1000

    records.update_all(file_number: new)
  end

  [la, av].each do |records|
    fail "Something looks wrong" if records.count > 1000

    records.update_all(vbms_id: new)
  end


  fail "Should just be one vet" if v.count > 1
  if vthing == :delete
    v.last.delete
  else
    v.last.update!()
  end
end

puts "Appeals updated: #{a.count}"
puts "Available Hearing Locations updated: #{l.count}"
puts "End Product Establishments updated: #{e.count}"
puts "Documents updated: #{d.count}"
puts "Intake updated: #{i.count}"
puts "HigherLevelReview updated: #{h.count}"
puts "SupplementalClaim updated: #{s.count}"
puts "Form8 updated: #{f.count}"
puts "RampElection updated: #{r.count}"
puts "RampRefiling updated: #{rr.count}"
puts "LegacyAppeal updated: #{la.count}"
puts "ApiView updated: #{av.count}"

if vthing == :delete
  puts "deleted veteran record id: #{v.last.id} in favor of id: #{newv.id}"
else
  puts "updated veteran record file number"
end
```

**Comment by pkarman** [comment]
Probably worth turning this into a rake task. I can't imagine this is the last time we'll do this.

**Comment by laurjpeterson** [comment]
Sally: we should try to fetch by file number first instead of assuming SSN is the file number.

**Comment by pkarman** [comment]
The way I read the code, we are always searching Caseflow db first for file number, and then if we can't find it, fetching the file number from BGS via SSN. So we should only ever be writing file numbers to the Caseflow db. @leikkisa can you point me at the code where that is not true?

**Comment by pkarman** [comment]
There appear to be zero records with ssn as filenumber in the db currently:

```
irb(main):002:0> Veteran.all.select { |v| v.file_number.to_i == v.file_number.to_s && v.file_number.length == 9 }.count
[2019-06-19 11:56:40 -0400]   Veteran Load (367.0ms)  SELECT "veterans".* FROM "veterans"
=> 0
```

**Comment by pkarman** [comment]
This query should find veterans in Caseflow with SSN-looking values in the `file_number` field:
```
Veteran.all.select { |v| v.file_number.to_i.to_s == v.file_number.to_s && v.file_number.length == 9 }
```
currently there are 191165

**Comment by pkarman** [comment]
fwiw, I tested BGS service and it will return the SSN as the file number: `BGSService.new.fetch_file_number_by_ssn`

**Comment by leikkisa** [comment]
I checked the last 1000 veterans, and 42 of them had the SSN stored as the file number, when the actual file number is different.  Often the filenumber and SSN are the same.  

The issue is that some BGS calls work with file number OR ssn, but not all of them. I think we had previously assumed that the calls that use filenumber don't also work by SSN.

Here are the Caseflow IDs of some of them where we saved them incorrectly.
[286461, 286507, 286593, 286609, 286620, 286621, 286633, 286651, 286660, 286702, 286733, 286753, 286773, 286804, 286926, 286973, 286987, 286996, 287012, 287036, 287087, 287131, 287169, 287180, 287214, 287227, 287228, 287232, 287234, 287268, 287288, 287297, 287317, 287320, 287361, 287366, 287383, 287410, 287413, 287419, 287448, 287457]

My suggestion would be to remove the first find_or_create_by_file_number in find_or_create_by_file_number_or_ssn.
